libftl/charon_protocol.c

```diff
@@ -52,3 +60,61 @@ ftl_charon_response_code_t ftl_charon_read_response_code(const char * response_s
    /* Got an invalid or unknown response code */
    return FTL_CHARON_UNKNOWN;
  }
+
+unsigned char decode_hex_char(char c) {
+    if (c >= '0' && c <= '9') {
+        return c - '0';
+    }
+    if (c >= 'A' && c <= 'Z') {
+        return (c - 'A') + 10;
+    }
+    if (c >= 'a' && c <= 'z') {
+        return (c - 'a') + 10;
+    }
+
+    return 0;
+}
+
+int recv_all(int sock, char * buf, int buflen) {
+    int pos = 0;
+    while (pos == 0 || buf[pos - 1] != '\n') {
+        pos += recv(sock, buf + pos, buflen - pos, 0);
```

Comment:
in addition to what @connor4312,  recv_all should return -1 if recv returns -1 to be consistent with recv that way getlasterror() will be invoked. if the connection was closed recv will also return 0 which will cause an infinite loop.  also light-styx seems to be inconsistent in it's line endings (which needs to be fixed on light-styx) and many of the errors are not terminated with a \n eg:  c.conn.Write([]byte("401 Unauthorized. Channel not found"))